### PR TITLE
Implement move-via-deprecate of `decToMaybe` as per #2330

### PR DESCRIPTION
**Pull request by JacquesCarette**
Make the main home of `decToMaybe` be `Relation.Nullary.Decidable.Core`, but
deprecate the one in `Data.Maybe.Base` instead of removing it entirely. Fix the library accordingly.

Note that this forces `Relation.Nullary.Decidable.Core` to use `Agda.Builtin.Maybe` to avoid a cyclic import. This can be fixed once the deprecation is done.

### CODE REVIEW

**Comment by jamesmckinna** [comment]
Needs a `CHANGELOG`!

**Comment by JacquesCarette** [comment]
Indeed it does.

**Comment by jamesmckinna** [comment]
Also, is this an opportunity to follow through with @gallais 's suggestions following [his comment](https://github.com/agda/agda-stdlib/issues/2059#issuecomment-1826775133) on #2059 ? Given that once you've deprecated, you get to pick new/better names and change the API as well? or would you rather leave this to a subsequent PR? I think it would be a good idea... on whatever basis.

**Comment by JacquesCarette** [comment]
`isTrue` seems sensible.

**Comment by jamesmckinna** [comment]
My preference, now that it's under `Decidable`, would be #2059 's `dec⇒weaklyDec`... because all of those uses under `Solver` are about weakly deciding a property, but are written under the identification of that with `Maybe`...

... but `isBlah` is at least a conventional use. (Style guide again? Is it documented anywhere?)

I wonder if `isTrue` is a bit on the nose? (Potential conceptual clash with `True` denoting `Bool`?)

I'm really glad to be revisiting some of the choices under #2055 / #2058 / #2059 which I suspect got lost in the wash of the v2.0 release (and my mistake in tagging those as v3.0...), but it feels as though this more modest PR might be leaving downstream hostages to fortune.

Perhaps, against myself, the thing is to stick with `decToMaybe` ahead of wider consideration of the naming issues. I feel genuinely conflicted!

File: src/Relation/Nullary/Decidable/Core.agda
Comment (gallais) [inline src/Relation/Nullary/Decidable/Core.agda:15]:
```suggestion
-- decToMaybe was deprecated in v2.1
-- this can go through `Data.Maybe.Base` once the deprecation of `decToMaybe`
-- is fully done.
```
Let's reuse the magic string to make greping for things that need to be removed easier.

File: src/Data/Maybe/Base.agda
Comment (jamesmckinna) [inline src/Data/Maybe/Base.agda:151]: Why not simply give the full qualified name `Relation.Nullary.Decidable.decToMaybe` instead?

**Review by jamesmckinna — APPROVED** [review]
Having blundered in with my own commits, revising @JacquesCarette 's original PR, it's perhaps out-of-order to give an 'Approve' review... but there you have it.

**Comment by JacquesCarette** [comment]
Until I hear at least one more opinion, I'm going to leave it as `decToMaybe`. I do agree its name should change!

File: src/Data/Maybe/Base.agda
Comment (MatthewDaggitt) [inline src/Data/Maybe/Base.agda:142]: To be honest, I think the easiest approach which would avoid all the import cycle comments would just be to entirely redefine it here. It's pretty short...

**Review by MatthewDaggitt — COMMENTED** [review]
To be honest if we are going to change the name, I think it would be good to change it now! My main reason is that it will enhance the backwards compatbility of this change. Currently if users have both `Data.Maybe` and `Relation.Nullary` open (which is entirely possible), then they'll get a duplicate name clash.

I don't have a strong opinion on the name.

File: src/Data/Maybe/Base.agda
Comment (JacquesCarette) [reply src/Data/Maybe/Base.agda:142]: Sure, can do.

**Comment by JacquesCarette** [comment]
I'm fine with renaming it, and `isYes` is indeed not the best name either. But I must say that I hate `dec⇒weaklyDec`. The `dec⇒` part is fine, it's the `weaklyDec` that I'm not happy with. What the heck is a "weak decision"? If this was long established vocabulary, sure. But I don't think that's the case. And 'weak' is ridiculously overloaded (and under-indicative).

In some ways `justYes` might be better (no, I'm not trying to pun here!) I would be fine with `dec⇒justYes`. I'm using 'just' in its 'truncate' sense here. I'm explicitly avoiding the term 'truncate' in the name, as this is a partial truncation, as that's a worse can of worms to open, IMHO.

**Comment by jamesmckinna** [comment]
I'd cheerfully take responsibility for the bad name, had I introduced it, but it seems to have been written into the library a long time ago.  

I sympathise with whoever first chose "weakly"  though: Johnstone in Topos Theory (1977) called `1+_` the "decidable partial map classifier" and there seems no easy way to extract a reasonable qualified adjective out of that, given that the other ready-at-hand prefixes "semi-" and "partial-" both connote the more established recursion theoretic 'yes definitely but may not terminate if no'... and recursion theory doesn't seem to have an agreed orthodox name for this useful middle ground? 

But v3.0 as a holding pen for reconsiderations of this name, and v2.x for continuing with it for the time being?

That said, the writers of the various `Solver`s seemed to bypass the issue entirely by working with `Maybe` directly, which is a different kind of (category) mistake IMHO (one of 'data vs. property'), so `WeaklyDecidable` has not really caught on. 

I think for the sake of this PR, and its focus on simplifying dependencies, suggest sticking with the name, even if you hate it?

File: src/Algebra/Solver/IdempotentCommutativeMonoid.agda
Comment (jamesmckinna) [inline src/Algebra/Solver/IdempotentCommutativeMonoid.agda:29]: For consistency with `Algebra.Solver.Monoid`, which uses `Dec.decToMaybe` qualified, maybe stick with the old `import` but with the qualified name `Dec`? or else expose `decToMaybe` at top-level in `...Monoid`?

File: src/Data/List/Relation/Binary/Sublist/Heterogeneous/Solver.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Binary/Sublist/Heterogeneous/Solver.agda:37]: Similarly to above...

**Comment by jamesmckinna** [comment]
I took the liberty of resolving the merge conflict, but without committing one way or the other to my suggestions regarding qualified names above.

File: src/Data/Maybe/Base.agda
Comment (JacquesCarette) [reply src/Data/Maybe/Base.agda:142]: Doesn't actually help: the direct definition still need `Relation.Nullary.Decidable.Core` to be imported.

**Comment by jamesmckinna** [comment]
Continues to look good!
Can we get a second review, please?

**Comment by jamesmckinna** [comment]
> Meeting conclusion:
> 
>     * Either change deprecation strategy to public re-export `Nullary.Core` name instead of redefining.
> 
>     * Or rename `Nullary.Core` definition

Suggest doing the first for v2.1 (hence, with no deprecation, IIUC), and deciding on the second (with consequent deprecation) as a downstream v2.2?

**Comment by jamesmckinna** [comment]
Apologies! the last commit appears to have broken things, because `Relation.Binary.Consequences`, *despite* importing form the 'right' place, nevertheless triggers the deprecation warning...

**Comment by JacquesCarette** [comment]
Implemented @MatthewDaggitt 's first deprecation strategy. This should be ready to go now.

**Comment by JacquesCarette** [comment]
Yes, that's my thinking.

**Comment by jamesmckinna** [comment]
> Yes, that's my thinking.

Happy to push this if your hands are full!

**Comment by jamesmckinna** [comment]
> I'm fine with isYes and isNo.

Aaargh! Wasn't thinking...

... `isYes` is the name of the projection from `Dec` to `Bool` %!$#

**Comment by jamesmckinna** [comment]
Moved name discussion back to the issue... sigh.

**Comment by MatthewDaggitt** [comment]
Yes, as I said in the meeting, you can't directly attach deprecation warnings to re-exports. As @jamesmckinna suggests we'd have to remove the deprecation warning entirely and leave just a textual comment that it is deprecated. This is what we've done in the past in similar circumstances.

**Comment by jamesmckinna** [comment]
Looking at the details of this PR, much of the knock-on viscosity arising from any name change to `decToMaybe` could be dealt with by fixing the various DRY violations under `Algebra.Solver.*Monoid`, and by delegation from a given `Relation.Binary.Definitions.DecidableEquality` via `Relation.Binary.Consequences.dec⇒weaklyDec`... (and we can fix *that* name downstream...)

**Comment by JacquesCarette** [comment]
Yes, this works. I'm fine with `dec⇒maybe`. Thanks.

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**
